sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlockBlobAsyncClient.java

```diff
@@ -156,6 +159,150 @@
+     *         this method will be and determines the number of requests that need to be made. The amount of memory
+     *         consumed by this method may be up to blockSize * numBuffers. If block size is large, this method will
+     *         make fewer network calls, but each individual call will send more data and will therefore take longer.
+     * @param numBuffers The maximum number of buffers this method should allocate. Must be at least two. Typically, the larger
+     *         the number of buffers, the more parallel, and thus faster, the upload portion of this operation will be.
+     *         The amount of memory consumed by this method may be up to blockSize * numBuffers.
+     * @param headers {@link BlobHTTPHeaders}
+     * @param metadata {@link Metadata}
+     * @param accessConditions {@link BlobAccessConditions}
+     * @return A reactive response containing the information of the uploaded block blob.
+     */
+    public Mono<Response<BlockBlobItem>> upload(Flux<ByteBuffer> data, int blockSize, int numBuffers,
+        BlobHTTPHeaders headers, Metadata metadata, BlobAccessConditions accessConditions) {
+        // TODO: Parallelism parameter? Or let Reactor handle it?
+        // TODO: Sample/api reference
+        Objects.requireNonNull(data, "data must not be null");
+        BlobAccessConditions accessConditionsFinal = accessConditions == null
+            ? new BlobAccessConditions() : accessConditions;
+
+        // TODO: Progress reporting.
+        // See ProgressReporter for an explanation on why this lock is necessary and why we use AtomicLong.
+        /*AtomicLong totalProgress = new AtomicLong(0);
+        Lock progressLock = new ReentrantLock();*/
+
+        // Validation done in the constructor.
+        UploadBufferPool pool = new UploadBufferPool(numBuffers, blockSize);
+
+        /*
+        Break the source Flux into chunks that are <= chunk size. This makes filling the pooled buffers much easier
+        as we can guarantee we only need at most two buffers for any call to write (two in the case of one pool buffer
+        filling up with more data to write). We use flatMapSequential because we need to guarantee we preserve the
+        ordering of the buffers, but we don't really care if one is split before another.
+         */
+        Flux<ByteBuffer> chunkedSource = data.flatMapSequential(buffer -> {
+            if (buffer.remaining() <= blockSize) {
+                return Flux.just(buffer);
+            }
+            List<ByteBuffer> smallerChunks = new ArrayList<>();
+            int numSplits = (int) Math.ceil(buffer.remaining() / (double) blockSize);
+            for (int i = 0; i < numSplits; i++) {
+                // Note that duplicate does not duplicate data. It simply creates a duplicate view of the data.
+                ByteBuffer duplicate = buffer.duplicate();
```

Comment:
I haven't been concerned about it because the network layer we're sending it to shouldn't be changing the data, and I don't think there's any more risk here of the data being transferred getting modified than in the other upload methods. Unless I'm missing something?

Reply:
Nope, was just asking.